### ISSUE TITLE
Auto complete doesn't work with corporate proxy

### ISSUE THREAD

**Issue by uxkjaer** [SAP/ui5-language-assistant#285]
Hi,

Excellent to see SAP taking ownership of these extensions. However the extension doesn't work through corporate proxy, even though it has been added in VS code settings and windows environment variables.

See output below

c:\Users\KjaerJ\AppData\Roaming\Code\User\globalStorage\saposs.vscode-ui5-language-assistant\ui5-resources-cache\1.71.14 will be used to cache UI5 resources
[Error - 9:46:22 am] Request textDocument/hover failed.
Message: Request textDocument/hover failed with message: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
Code: -32603
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:13616) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:13616) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
[Error - 9:46:31 am] Request textDocument/hover failed.
Message: Request textDocument/hover failed with message: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
Code: -32603
[Error - 9:46:32 am] Request textDocument/completion failed.
Message: Request textDocument/completion failed with message: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
Code: -32603
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
[Error - 9:48:13 am] Request textDocument/hover failed.
Message: Request textDocument/hover failed with message: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
Code: -32603
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 4)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 4)
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 5)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 5)
[Error - 9:48:13 am] Request textDocument/completion failed.
Message: Request textDocument/completion failed with message: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
Code: -32603
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: FetchError: request to https://sapui5.hana.ondemand.com/1.71.14/test-resources/sap/ca/scfld/md/designtime/api.json failed, reason: connect ETIMEDOUT 23.48.30.95:443
at ClientRequest. (c:\Users\KjaerJ.vscode\extensions\saposs.vscode-ui5-language-assistant-1.5.1\node_modules@ui5-language-assistant\language-server\dist\server.js:48289:11)
at ClientRequest.emit (events.js:203:13)
at TLSSocket.socketErrorListener (_http_client.js:399:9)
at TLSSocket.emit (events.js:203:13)
at emitErrorNT (internal/streams/destroy.js:91:8)
at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
at processTicksAndRejections (internal/process/task_queues.js:77:11)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 6)
(node:13616) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 6)

**Comment by bd82** [comment]
We are using `node-fetch` to make the http requests to retrieve the UI5 SDK.

It seems that to support proxies a custom `agent` need to be passed.
- https://github.com/node-fetch/node-fetch#custom-agent
- https://github.com/node-fetch/node-fetch/issues/892

So the question is if we should created a custom proxyAgent conditionally if certain env varialbes
are defined or replace `node-fetch` with some other http requests library which does that auto-magically?

**Comment by bd82** [comment]
`Axios` library configuration seems to imply that it would auto-magically respect proxy related env variables.
- https://github.com/axios/axios#request-config (scroll down until `proxy` key).

Using it may be optimal instead of adding a branch to our code and having to respect multiple
env variables (HTTP_PROXY, HTTPS_PROXY, NO_PROXY, auth related env vars?)

**Comment by bd82** [comment]
@uxkjaer this is not a high priority for us as our main scenarios do not include a corporate proxy.
Also I don't actually have an env with a corporate proxy to test changes on.

**Would you be interested in attempting to contribute a PR for this issue?**

This is the relevant file responsible for fetching the UI5 SDK.
- https://github.com/SAP/ui5-language-assistant/blob/master/packages/language-server/src/ui5-model.ts

**Comment by uxkjaer** [comment]
Hi @bd82 

I've forked the repo and are trying to experiment with this. Bear with me, it's my first time doing typescript. I've cloned the repo down, but can't even run the tests as the I get this error:
lerna ERR! yarn run test exited 1 in '@ui5-language-assistant/settings'
lerna ERR! yarn run test stdout:
yarn run v1.22.4
$ mocha
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

lerna ERR! yarn run test stderr:
Error: No test files found: "./test/**/*spec.js"
error Command failed with exit code 1.

I created a small test program which shows what would need to be done from our end to make this work. You can check the repo here:
https://github.com/uxkjaer/ui5_lang_serv_test

**Comment by bd82** [comment]
You need to compile the code first before you can run the tests.
Try running `yarn run compile:watch` at the top level of the repo and leave that processes running
in a separate terminal before running any test scripts.

**Comment by bd82** [comment]
Also if you run `yarn run build:quick` at the top level of the package
it would build a new "*.VSIX" file in the `packages/vscode-ui5-language-assistant` 
that you can install in your local VSCode

**Comment by ndrsg** [comment]
I tried to provide a fix and created a PR, but i cannot fulfill the lint requirements due a issue with agent-base for http-proxy-agent and https-proxy agent. (Type incompatibility)

https://github.com/TooTallNate/node-https-proxy-agent/issues/108

**Comment by bd82** [comment]
Thanks @ndrsg 👍 

I will have a look at your PR.
I believe we can disable the linter for a specific line using an "ignore comment"

**Comment by uxkjaer** [comment]
how's it going with this issue. It's stopping us for using this excellent extension. I believe many would be in the same boat.

**Comment by bd82** [comment]
Hello @uxkjaer. 

The Open PR (#302) seems to have been abandoned...

**Comment by babuilyas** [comment]
I'm tired of work arounds with corp proxy. @uxkjaer any progress on using this extension. I try add https://www.npmjs.com/package/node-global-proxy into the extension activate and it failed to get it working.

**Comment by bd82** [comment]
Hello @babuilyas 

Do you want to try taking over the #302 PR? 

Cheers.
Shahar.

**Comment by babuilyas** [comment]
@bd82 I think, i will be in position to do so in few days. Getting up with ui5 tooling and fiori tools.

**Comment by bd82** [comment]
Great @babuilyas 

Feel free to ping me if you have any questions.

**Comment by babuilyas** [comment]
> Great @babuilyas 👍
> 
> Feel free to ping me if you have any questions.

Thanks. Check it out.

https://github.com/SAP/ui5-language-assistant/pull/302#issuecomment-737253345

**Comment by bd82** [comment]
Thanks @babuilyas 👍 

I am re-posting your comment here for future reference:

```
I found a hack resolved this for me in vscode while working on other related issue.
My dev machine runs on Windows, so adopt appropriate your OS.
Windows Shell and Power Shell uses same proxy, so setup it with following command

Step 1: C:>netsh winhttp set proxy http://172.16.0.38:808 bypass-list="localhost;127.0.0.1"
[Requires Admin rights]

Step 2: Launch VSCode from command with
C:\your-ui5-project> code .

Step 3: Optional confirm proxy in VScode terminal with
C:\your-ui5-project> netsh winhttp show proxy
Result: Step 1 proxy output

Step 4: Confirm proxy is working in VScode terminal with
C:\your-ui5-project> curl "https://www.google.com"
Result: StatusCode : 200
StatusDescription : OK

Now try code-completion and inline documentation etc.
It works for me 👍 and should work for you.

Credit goes to @petermuessig
```

**Comment by zutos** [comment]
Hey, just wondering if this is still a valid work around? I've tried those steps and confirmed the proxy is working but the extension still throws the connection errors from the OP. This extension seems really useful for my team but this is preventing us :/

**Comment by babuilyas** [comment]
working behind proxy always sucks, so be sure your proxy is used by vs code as well. Try any proxy app like some mentioned in this post.
https://www.raymond.cc/blog/route-all-internet-software-and-game-connection-through-open-proxy-servers/

**Comment by uxkjaer** [comment]
I'd like to take another stab at this issue. I've been trying to get the source working so I can debug. However I'm getting TS errors left right and center. Can you provide a readme of how to get it running from the source code to be able to debug and fix.

**Comment by bd82** [comment]
Hello @uxkjaer I've replied to your ts-compilation question in the other issue you opened
- #433

**Comment by uxkjaer** [comment]
hi @bd82,
I've created a draft merge request that i'm working on. This will replace node-fetch with axios and also handle the corporate proxy. Currently all tests are passing in the language server except one. The test 
"it("does not cache the model"
fails because i'm not using fetcher anymore. A little help would be appreciated.

**Comment by bd82** [comment]
@uxkjaer it could be a while before I will have any time to look into this....

**Comment by Fjaoos** [comment]
Hi @all,

first I want to thank @uxkjaer for his work. I am currently trying to evaluate if the company I am working for should switch to BAS or VSC but without language support I cannot recommend VSCode.

The hack did not work for me as well and most developers won't have adminstrative privileges anyway.

Is it possible to use a local file?

**Comment by SharadaSowmya14** [comment]
Hi All,
The hack specified in #302 is not working for me as well. May I know if there is a possibility to have the fix for this issue in the future releases?

**Comment by uxkjaer** [comment]
Hi Sharada,

You can try to clone my fork down and build that. This is working for me,
currently waiting on a pull request to complete.

Kind regards
Jakob Marius Kjær
PHone: (+61) 0413 980 556

Linkedin: http://dk.linkedin.com/in/jmkjaer
Twitter: https://twitter.com/uxkjaer

**Comment by SharadaSowmya14** [comment]
Thank you @uxkjaer! I am able to use the extension now.

**Comment by sammok83** [comment]
Hi all, 
whats the status on this?
FYI, currently using VSCode Portable with extensions preloaded first as a workaround

**Comment by uxkjaer** [comment]
> Hi all, 
> whats the current status on this?
> FYI, currently using VSCode Portable with extensions preloaded first as a workaround

Check out my pr #435 or clone my repo directly and build to get the vsix. Then install from that. This works with proxy.

https://github.com/uxkjaer/ui5-language-assistant

**Comment by sammok83** [comment]
> > Hi all,
> > whats the current status on this?
> > FYI, currently using VSCode Portable with extensions preloaded first as a workaround
> 
> Check out my pr #435 or clone my repo directly and build to get the vsix. Then install from that. This works with proxy.
> 
> https://github.com/uxkjaer/ui5-language-assistant

Thanks @uxkjaer, but it's not working for me.

Tried:
1. Running the build:quick script
2. The .vsix file was generated 
3. Uninstalling the latest UI5 Language Assistant extension and installing the .vsix file in step 2.

But it didn't work. Any steps I may have missed?

**Comment by uxkjaer** [comment]
Do you have your proxy settings added as environment variables?

**Comment by sammok83** [comment]
Yes, as windows env variables and even tried it as VS Code level to overwrite it

**Comment by francoisvl** [comment]
> > Hi all,
> > whats the current status on this?
> > FYI, currently using VSCode Portable with extensions preloaded first as a workaround
> 
> Check out my pr #435 or clone my repo directly and build to get the vsix. Then install from that. This works with proxy.
> 
> https://github.com/uxkjaer/ui5-language-assistant

The current official version still does not work from behind a proxy for us, so the @uxkjaer fork still seems to be the way to go. A small tweak was required to spec a newer hard-coded UI5 version, as SAP has started removing older libraries from their servers and the version specified in the current fork is no longer available. See our PR [#1](https://github.com/uxkjaer/ui5-language-assistant/pull/1) to Jakob's fork.

**Comment by petermuessig** [comment]
Related PR #475

**Comment by petermuessig** [comment]
New related PR is #478 (made a mistake 😒 )

**Comment by uxkjaer** [comment]
I tested this locally with fidder as a proxy and it worked, however spoken to a friend, who tested this at a government client and it's still failing. He installed the vsix from the PR #478 



![image](https://user-images.githubusercontent.com/8186370/186326521-72012520-4e74-44f8-abfa-0b10485513d8.png)

**Comment by petermuessig** [comment]
Check this out: [vscode-ui5-language-assistant-3.2.1-proxy.vsix.zip](https://github.com/SAP/ui5-language-assistant/files/9413229/vscode-ui5-language-assistant-3.2.1-proxy.vsix.zip)

**Comment by petermuessig** [comment]
Should be fixed with #483 - waiting for confirmation

**Comment by petermuessig** [comment]
Confirmed with local proxy tests using *HTTP Toolkit*
